internal/resources/resource_with_unexported_fields.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2022 VMware, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/vmware-labs/reconciler-runtime/apis"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var (
+	_ webhook.Defaulter = &TestResourceUnexportedFields{}
+	_ webhook.Validator = &TestResourceUnexportedFields{}
+	_ client.Object     = &TestResourceUnexportedFields{}
+)
+
+// +kubebuilder:object:root=true
+// +genclient
+
+type TestResourceUnexportedFields struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TestResourceUnexportedFieldsSpec   `json:"spec"`
+	Status TestResourceUnexportedFieldsStatus `json:"status"`
+}
+
+func (r *TestResourceUnexportedFields) Default() {
+	if r.Spec.Fields == nil {
+		r.Spec.Fields = map[string]string{}
+	}
+	r.Spec.Fields["Defaulter"] = "ran"
+}
+
+func (r *TestResourceUnexportedFields) ValidateCreate() (admission.Warnings, error) {
+	return nil, r.validate().ToAggregate()
+}
+
+func (r *TestResourceUnexportedFields) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	return nil, r.validate().ToAggregate()
+}
+
+func (r *TestResourceUnexportedFields) ValidateDelete() (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (r *TestResourceUnexportedFields) validate() field.ErrorList {
+	errs := field.ErrorList{}
+
+	if r.Spec.Fields != nil {
+		if _, ok := r.Spec.Fields["invalid"]; ok {
+			field.Invalid(field.NewPath("spec", "fields", "invalid"), r.Spec.Fields["invalid"], "")
+		}
+	}
+
+	return errs
+}
+
+func (r *TestResourceUnexportedFields) CopyUnexportedFields() {
+	r.Status.unexportedFields = r.Spec.unexportedFields
+}
+
+// +kubebuilder:object:generate=true
+type TestResourceUnexportedFieldsSpec struct {
+	Fields           map[string]string `json:"fields,omitempty"`
+	unexportedFields map[string]string
```

Comment:
it wasn't clear to me whether the cmp IgnoreUnexported option works recursively or for the top level struct only. IMHO, it'd be useful to have a nested struct with unexported fields as part of this spec.

Reply:
I see you have implemeted your own opt. Does it handle nested structs with unexported fields?

Reply:
Within cmp, each option is applied to every path discovered while walking the object tree. If the path is filtered no fields under it are processed.